### PR TITLE
Setting up Model Tests

### PR DESCRIPTION
**Pull request by mccullagh**
First Unit Test for **DeicesModel**, sqlite::memory: is used as DB.

for #17

### CODE REVIEW

**Comment by mccullagh** [comment]
I will make the changes tomorrow. But @dbudwin can you give me feedback if that is what you imagined?

**Comment by dbudwin** [comment]
Sure, I'll look it over.  I don't claim to be a PHP or PHPUnit expert, but I'll provide what feedback I can.

File: tests/phpunit.xml [hunk 1]
```diff
@@ -7,4 +7,10 @@
     <testsuite>
         <directory>./</directory>
     </testsuite>
+    <php>
+        <var name="DB_DSN" value="sqlite::memory:" />
+        <var name="DB_USER" value="" />
+        <var name="DB_PASSWD" value="" />
```
Comment (dbudwin) [inline tests/phpunit.xml]: Please name this to `DB_PASSWORD`

File: tests/phpunit.xml [hunk 2]
```diff
@@ -7,4 +7,10 @@
     <testsuite>
         <directory>./</directory>
     </testsuite>
+    <php>
+        <var name="DB_DSN" value="sqlite::memory:" />
```
Comment (dbudwin) [inline tests/phpunit.xml]: I'm guessing "DSN" stands for "Database Server Name?"  Please rename to `DB_SERVER` or something a bit more descriptive.

File: tests/phpunit.xml [hunk 3]
```diff
@@ -7,4 +7,10 @@
...
+        <var name="DB_USER" value="" />
+        <var name="DB_PASSWD" value="" />
+        <var name="DB_DBNAME" value="" />
```
Comment (dbudwin) [inline tests/phpunit.xml]: Please rename to `DB_NAME`.

**Comment by dbudwin** [comment]
Hey @mccullagh just checking in.  Can I help you with anything related to this PR?

**Comment by dbudwin** [comment]
Hello @mccullagh, do you have any plans to continue working on this?  I know life can get in the way sometimes, but if you can't get to it in the next day or two, I'll take this PR over to keep things moving.

**Comment by mccullagh** [comment]
Hi @dbudwin , sorry that i took so long. Tomorrow i can continue to work on it.

**Comment by dbudwin** [comment]
No worries @mccullagh!  I've learned a little bit about PHPUnit and I'm going to update my feedback since some of it isn't applicable anymore.

File: tests/app/models/DevicesModelTest.php [hunk 4]
```diff
@@ -0,0 +1,38 @@
+<?php
+
+use Models\DevicesModel;
+
+class DeicesModelTest extends BaseModelTest
+{
+    public function testConstruct()
```
Comment (dbudwin) [inline tests/app/models/DevicesModelTest.php]: This doesn't seem to fulfill a purpose.  All this does is declare a local variable.  Is there a reason for this?

Comment (mccullagh) [inline tests/app/models/DevicesModelTest.php]: Idee was as a smoke test.

File: tests/app/models/DevicesModelTest.php [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    {
+        $deviceModel = new DevicesModel($this->getConnection());
+    }
+
+    public function testAdd()
+    {
+        $_POST = array('Name' => 'dummy', 'Description' => 'lorem ipsum', 'Type' => 1);
+
+        $deviceModel = new DevicesModel($this->getConnection());
```
Comment (dbudwin) [inline tests/app/models/DevicesModelTest.php]: `setUp()` should give you a fresh instance before each test is run.  `setUp()` is automatically called before each individual test is run and is a good place to put code that is shared among many tests.

https://phpunit.de/manual/current/en/fixtures.html

File: tests/phpunit.xml [hunk 6]
```diff
@@ -7,4 +7,10 @@
     <testsuite>
         <directory>./</directory>
     </testsuite>
+    <php>
+        <var name="DB_DSN" value="sqlite::memory:" />
+        <var name="DB_USER" value="" />
+        <var name="DB_PASSWORD" value="" />
+        <var name="DB_DBNAME" value="" />
```
Comment (dbudwin) [inline tests/phpunit.xml]: Please rename to remove the redundant `DB`: `DB_NAME`.

File: tests/app/models/BaseModelTest.php [hunk 7]
```diff
@@ -0,0 +1,21 @@
+<?php
+
+use PHPUnit\Framework\TestCase;
+use DB\SQL;
+
+abstract class BaseModelTest extends TestCase
+{
+
```
Comment (dbudwin) [inline tests/app/models/BaseModelTest.php]: Please remove the extra new line here.

Over all this class looks much sleeker now!  Good work!

File: tests/app/models/BaseModelTest.php [hunk 8]
```diff
@@ -0,0 +1,21 @@
+<?php
+
+use PHPUnit\Framework\TestCase;
+use DB\SQL;
+
```
Comment (dbudwin) [inline tests/app/models/BaseModelTest.php]: Codeacy is wanting a namespace for these new classes.  It seems like the convention for PHP and PHPUnit is to use the same namespace as the class being tested.  How do you feel about this?

http://stackoverflow.com/questions/12117254/organizing-phpunit-tests-in-namespaces

File: tests/app/models/DevicesModelTest.php [hunk 9]
```diff
@@ -0,0 +1,32 @@
+<?php
+
+use Models\DevicesModel;
+
+class DevicesModelTest extends BaseModelTest
+{
+
+    public function testAdd()
```
Comment (dbudwin) [inline tests/app/models/DevicesModelTest.php]: Ignore this.  I don't like the Codacy bot, just listen to Code Climate :)

**Comment by mccullagh** [comment]
@dbudwin i don't know why the autoloader doesn't work in PHP 7.1 anymore.

**Comment by dbudwin** [comment]
Looking on Google, I see a lot of `composer.json` files that specify a `psr` field.  Could it be related to that? https://seld.be/notes/psr-4-autoloading-support-in-composer

**Comment by dbudwin** [comment]
Ooo. I have an idea!  Try rebasing onto `master`!  Since you opened this PR I changed the case of the folders due to autoloading issues from `models`, `views`, `controllers`, to `Models`, `Views`, `Controllers`.  You may need to change the name of the `models` folder in `test` to `Models`.  See https://github.com/dbudwin/RoboHome-Web/pull/35

Also, for PHP 5.6 and 7.0...even though it says it passed on Travis, there is actually an error and none of the tests are being run: "file_get_contents(./schema.sql): failed to open stream: No such file or directory"

**Comment by dbudwin** [comment]
I pulled your branch and played with it some, I found a few things that may help.

1. In `BaseModelTest` the value of `$GLOBALS['DB_DSN']` is empty which is a problem.  When I hard coded the value it got me past the `PDOException: invalid data source name` error.  I'm not sure why it's unable to load the value from `phpunit.xml`.
2. Fixing the previous error exposed the next error.  I moved the location of the `schema.sql` file in a previous PR.  It's not in an ideal location right now but it's outside the scope of this PR, it should be changed to: `./docker/mysql/01schema.sql`.

This got me further, but it's still having trouble executing the tests.  It complains: `Undefined field ID`.  I'm guessing because it's not actually performing any actions on a database.

File: composer.json [hunk 10]
```diff
@@ -1,4 +1,11 @@
 {
+    "autoload": {
+        "files": ["fatfree/lib/base.php"],
+        "psr-4": {
+            "Models\\":"fatfree/app/models",
```
Comment (dbudwin) [inline composer.json]: This may need to be `"Models\\":"fatfree/app/Models"`.  Note the capital "M" in models.

File: composer.json [hunk 11]
```diff
@@ -1,4 +1,11 @@
...
+            "Controller\\":"fatfree/app/controllers"
```
Comment (dbudwin) [inline composer.json]: This may need to be `"Models\\":"fatfree/app/Controllers"`.  Note the capital "C" in controllers.

**Comment by dbudwin** [comment]
For testing, I replaced `'sqlite::memory:'` with `sqlite::test.db'` to look at the database manually, but the file is empty.  It's not actually loading the schema for some reason.  The `file_get_contents` reads the file, but the `exec` command doesn't seem to be working as expected.

**Comment by dbudwin** [comment]
Good news @mccullagh, I got the tests working!!!  It was tricky, but you gave me a good start with your previous commits.  I would appreciate if you could review my changes and provide any feedback in the next day or two.  If all looks good I can go ahead and merge it.

**Pull request merged**